### PR TITLE
New test project generator

### PR DESCRIPTION
**Pull request by jjohannes**

### CODE REVIEW

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorRunner.groovy [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+package org.gradle.performance.generator
+
+class TestProjectGeneratorRunner {
+
+    static void main(String[] args) {
+        def projectName = args[0]
+        def outputDir = new File(args[1])
+
+        JavaTestProject project = JavaTestProject.values().find { it.projectName == projectName }
```
Comment (oehme) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorRunner.groovy]: This is fine for now, but if we want to make this reusable then we can't have a hardcoded enum, but actually need to take the configuration from the command line or a file.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorRunner.groovy [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.performance.generator
+
+class TestProjectGeneratorRunner {
+
+    static void main(String[] args) {
```
Comment (oehme) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorRunner.groovy]: I think this main method can be part of TestProjectGenerator, so we don't need a "GeneratorRunner" :)

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/DependencyTree.groovy [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.performance.generator
+
+import org.gradle.internal.Pair
+
+class DependencyTree {
```
Comment (oehme) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/DependencyTree.groovy]: I'm a bit lost at what this class does. Some class documentation about the problem it solves would help review it.

Also I'm curious: We talked about creating dependencies purely based on simple arithmetic like "class n depends on class n - 3", so I'm not quite sure why we need to hold state at all?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        configurations {
+            compile.extendsFrom projectsConfiguration
+        }
+
+        dependencies {
+            ${config.externalApiDependencies.collect { "api '$it'" }.join("\n            ")}
+            ${config.externalImplementationDependencies.collect { "implementation '$it'" }.join("\n            ")}
+            testImplementation '${config.useTestNG ? 'org.testng:testng:6.4' : 'junit:junit:4.12'}'
+
+            ${subProjectNumber == null ? '' : dependencyTree.parentToChildrenNodeSets.get(subProjectNumber).collect { "${it % 9 == 0 ? 'api' : 'implementation'} project(':project$it')" }.join("\n            ")}
```
Comment (oehme) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy]: implementation/api only work with >3.4. We need to use `compile` for older versions.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        """
+    }
+
+    def generateGradleProperties(boolean isRoot) {
+        if (!isRoot) {
+            return null
+        }
+        """
+        org.gradle.jvmargs=-Xmxs${config.daemonMemory} -Xmx${config.daemonMemory}
+        org.gradle.parallel=${config.subProjects > 0}
```
Comment (oehme) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy]: Should be an explicit parameter

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorRunner.groovy (hunk 2 above)
Comment (jjohannes) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorRunner.groovy]: Done :)

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy (hunk 4 above)
Comment (jjohannes) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy]: Fixed now inside the generated `build.gradle`:
```
        def missingJavaLibrarySupport = GradleVersion.current() < GradleVersion.version('3.4')

        if (missingJavaLibrarySupport) {
            configurations {
                ${hasParent ? 'api' : ''}
                implementation
                testImplementation
                ${hasParent ? 'compile.extendsFrom api' : ''}
                compile.extendsFrom implementation
                testCompile.extendsFrom testImplementation
            }
        }
```

`hasParent` parent is used to decide whether it is a `java` or `java-library` project.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy (hunk 5 above)
Comment (jjohannes) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy]: Changed.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/DependencyTree.groovy (hunk 3 above)
Comment (jjohannes) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/DependencyTree.groovy]: I cleaned this up and documented it (at least a bit) now. 

This solution now also generates more "complex" class dependencies based on `api` project dependencies.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/JavaTestProject.groovy [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ */
+
+package org.gradle.performance.generator
+
+enum JavaTestProject {
+
+    LARGE_MONOLITHIC_JAVA_PROJECT("largeMonolithicJavaProject", 50000, 0, '4g'),
+    LARGE_JAVA_MULTI_PROJECT("largeJavaMultiProject", 100, 500, '256m'),
+
+    MEDIUM_MONOLITHIC_JAVA_PROJECT("mediumMonolithicJavaProject", 25000, 0, '4g'),
```
Comment (oehme) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/JavaTestProject.groovy]: I think we should reduce these to 100 projects, to make the term "medium" consistent with the TestNG test project below.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/experiment/maven/JavaTestGradleVsMavenPerformanceTest.groovy [hunk 7]
```diff
@@ -17,83 +17,70 @@
...
-        'mediumWithJUnit' | 'medium' | 'clean build and run tests' | ['clean', 'test']     | ['clean', 'test']    | 5000          | 100
-    }
-
-    @Unroll("Gradle vs Maven #description build for #template")
-    def "build performance test"() {
-        given:
-        runner.testGroup = "Gradle vs Maven build using Java plugin"
-        runner.testId = "$size $description with Java plugin"
-        equivalentGradleAndMavenBuilds(template, description, gradleTasks, equivalentMavenTasks)
+        runner.configure()
```
Comment (oehme) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/experiment/maven/JavaTestGradleVsMavenPerformanceTest.groovy]: The individual tests shouldn't have to call this. Either we make it the hard convention that there is always one Gradle and one Maven invocation that is configured implicitly through a few fields or we just add a convenience method that takes some parameters. Having a mix of fields and a no-arg method call will be confusing for the next person looking at that base class.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/JavaTestProject.groovy (hunk 6 above)
Comment (jjohannes) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/JavaTestProject.groovy]: Changed.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/experiment/maven/JavaTestGradleVsMavenPerformanceTest.groovy (hunk 7 above)
Comment (jjohannes) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/experiment/maven/JavaTestGradleVsMavenPerformanceTest.groovy]: Right. This is now done in run() without the need to call configure() first.

**Pull request merged**